README.md
# simple-react-full-stack

<!-- https://medium.com/@jpqo9508/building-a-simple-game-with-react-part-1-3d7df620678d
 -->

[![Build Status](https://travis-ci.org/crsandeep/simple-react-full-stack.svg?branch=master)](https://travis-ci.org/crsandeep/simple-react-full-stack)
[![Greenkeeper badge](https://badges.greenkeeper.io/crsandeep/simple-react-full-stack.svg)](https://greenkeeper.io/)

This is a boilerplate to build a full stack web application using React, Node.js, Express and Webpack. It is also configured with webpack-dev-server, eslint, prettier and babel.

- [simple-react-full-stack](#simple-react-full-stack)
  - [Introduction](#introduction)
    - [Development mode](#development-mode)
    - [Production mode](#production-mode)
  - [Quick Start](#quick-start)
  - [Documentation](#documentation)
    - [Folder Structure](#folder-structure)
    - [Babel](#babel)
    - [ESLint](#eslint)
    - [Webpack](#webpack)
    - [Webpack dev server](#webpack-dev-server)
    - [Nodemon](#nodemon)
    - [Express](#express)
    - [Concurrently](#concurrently)
    - [VSCode + ESLint + Prettier](#vscode--eslint--prettier)
      - [Installation guide](#installation-guide)

## Introduction

[Create React App](https://github.com/facebook/create-react-app) is a quick way to get started with React development and it requires no build configuration. But it completely hides the build config which makes it difficult to extend. It also requires some additional work to integrate it with an existing Node.js/Express backend application.

This is a simple full stack [React](https://reactjs.org/) application with a [Node.js](https://nodejs.org/en/) and [Express](https://expressjs.com/) backend. Client side code is written in React and the backend API is written using Express. This application is configured with [Airbnb's ESLint rules](https://github.com/airbnb/javascript) and formatted through [prettier](https://prettier.io/).

### Development mode

In the development mode, we will have 2 servers running. The front end code will be served by the [webpack dev server](https://webpack.js.org/configuration/dev-server/) which helps with hot and live reloading. The server side Express code will be served by a node server using [nodemon](https://nodemon.io/) which helps in automatically restarting the server whenever server side code changes.

### Production mode

In the production mode, we will have only 1 server running. All the client side code will be bundled into static files using webpack and it will be served by the Node.js/Express application.

## Quick Start

```bash
# Clone the repository
git clone https://github.com/crsandeep/simple-react-full-stack

# Go inside the directory
cd simple-react-full-stack

# Install dependencies
yarn (or npm install)

# Start development server
yarn dev (or npm run dev)

# Build for production
yarn build (or npm run build)

# Start production server
yarn start (or npm start)
```

## Documentation

### Folder Structure

All the source code will be inside **src** directory. Inside src, there is client and server directory. All the frontend code (react, css, js and any other assets) will be in client directory. Backend Node.js/Express code will be in the server directory.

### Babel

[Babel](https://babeljs.io/) helps us to write code in the latest version of JavaScript. If an environment does not support certain features natively, Babel will help us to compile those features down to a supported version. It also helps us to convert JSX to Javascript.

[.babelrc file](https://babeljs.io/docs/usage/babelrc/) is used describe the configurations required for Babel. Below is the .babelrc file which I am using.

```javascript
{
    "presets": ["env", "react"]
}
```

Babel requires plugins to do the transformation. Presets are the set of plugins defined by Babel. Preset **env** allows to use babel-preset-es2015, babel-preset-es2016, and babel-preset-es2017 and it will transform them to ES5. Preset **react** allows us to use JSX syntax and it will transform JSX to Javascript.

### ESLint

[ESLint](https://eslint.org/) is a pluggable and configurable linter tool for identifying and reporting on patterns in JavaScript.

[.eslintrc.json file](<(https://eslint.org/docs/user-guide/configuring)>) (alternatively configurations can we written in Javascript or YAML as well) is used describe the configurations required for ESLint. Below is the .eslintrc.json file which I am using.

```javascript
{
  "extends": ["airbnb"],
  "env": {
    "browser": true,
    "node": true
  },
  "rules": {
    "no-console": "off",
    "comma-dangle": "off",
    "react/jsx-filename-extension": "off"
  }
}
```

[I am using Airbnb's Javascript Style Guide](https://github.com/airbnb/javascript) which is used by many JavaScript developers worldwide. Since we are going to write both client (browser) and server side (Node.js) code, I am setting the **env** to browser and node. Optionally, we can override the Airbnb's configurations to suit our needs. I have turned off [**no-console**](https://eslint.org/docs/rules/no-console), [**comma-dangle**](https://eslint.org/docs/rules/comma-dangle) and [**react/jsx-filename-extension**](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md) rules.

### Webpack

[Webpack](https://webpack.js.org/) is a module bundler. Its main purpose is to bundle JavaScript files for usage in a browser.

[webpack.config.js](https://webpack.js.org/configuration/) file is used to describe the configurations required for webpack. Below is the webpack.config.js file which I am using.

```javascript
const path = require("path");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const CleanWebpackPlugin = require("clean-webpack-plugin");

const outputDirectory = "dist";

module.exports = {
  entry: ["babel-polyfill", "./src/client/index.js"],
  output: {
    path: path.join(__dirname, outputDirectory),
    filename: "bundle.js"
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader"
        }
      },
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"]
      },
      {
        test: /\.(png|woff|woff2|eot|ttf|svg)$/,
        loader: "url-loader?limit=100000"
      }
    ]
  },
  devServer: {
    port: 3000,
    open: true,
    proxy: {
      "/api": "http://localhost:8080"
    }
  },
  plugins: [
    new CleanWebpackPlugin([outputDirectory]),
    new HtmlWebpackPlugin({
      template: "./public/index.html",
      favicon: "./public/favicon.ico"
    })
  ]
};
```

1.  **entry:** entry: ./src/client/index.js is where the application starts executing and webpack starts bundling.
    Note: babel-polyfill is added to support async/await. Read more [here](https://babeljs.io/docs/en/babel-polyfill#usage-in-node-browserify-webpack).
2.  **output path and filename:** the target directory and the filename for the bundled output
3.  **module loaders:** Module loaders are transformations that are applied on the source code of a module. We pass all the js file through [babel-loader](https://github.com/babel/babel-loader) to transform JSX to Javascript. CSS files are passed through [css-loaders](https://github.com/webpack-contrib/css-loader) and [style-loaders](https://github.com/webpack-contrib/style-loader) to load and bundle CSS files. Fonts and images are loaded through url-loader.
4.  **Dev Server:** Configurations for the webpack-dev-server which will be described in coming section.
5.  **plugins:** [clean-webpack-plugin](https://github.com/johnagan/clean-webpack-plugin) is a webpack plugin to remove the build folder(s) before building. [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin) simplifies creation of HTML files to serve your webpack bundles. It loads the template (public/index.html) and injects the output bundle.

### Webpack dev server

[Webpack dev server](https://webpack.js.org/configuration/dev-server/) is used along with webpack. It provides a development server that provides live reloading for the client side code. This should be used for development only.

The devServer section of webpack.config.js contains the configuration required to run webpack-dev-server which is given below.

```javascript
devServer: {
    port: 3000,
    open: true,
    proxy: {
        "/api": "http://localhost:8080"
    }
}
```

[**Port**](https://webpack.js.org/configuration/dev-server/#devserver-port) specifies the Webpack dev server to listen on this particular port (3000 in this case). When [**open**](https://webpack.js.org/configuration/dev-server/#devserver-open) is set to true, it will automatically open the home page on startup. [Proxying](https://webpack.js.org/configuration/dev-server/#devserver-proxy) URLs can be useful when we have a separate API backend development server and we want to send API requests on the same domain. In our case, we have a Node.js/Express backend where we want to send the API requests to.

### Nodemon

Nodemon is a utility that will monitor for any changes in the server source code and it automatically restart the server. This is used in development only.

nodemon.json file is used to describe the configurations for Nodemon. Below is the nodemon.json file which I am using.

```javascript
{
  "watch": ["src/server/"]
}
```

Here, we tell nodemon to watch the files in the directory src/server where out server side code resides. Nodemon will restart the node server whenever a file under src/server directory is modified.

### Express

Express is a web application framework for Node.js. It is used to build our backend API's.

src/server/index.js is the entry point to the server application. Below is the src/server/index.js file

```javascript
const express = require("express");
const os = require("os");

const app = express();

app.use(express.static("dist"));
app.get("/api/getUsername", (req, res) =>
  res.send({ username: os.userInfo().username })
);
app.listen(8080, () => console.log("Listening on port 8080!"));
```

This starts a server and listens on port 8080 for connections. The app responds with `{username: <username>}` for requests to the URL (/api/getUsername). It is also configured to serve the static files from **dist** directory.

### Concurrently

[Concurrently](https://github.com/kimmobrunfeldt/concurrently) is used to run multiple commands concurrently. I am using it to run the webpack dev server and the backend node server concurrently in the development environment. Below are the npm/yarn script commands used.

```javascript
"client": "webpack-dev-server --mode development --devtool inline-source-map --hot",
"server": "nodemon src/server/index.js",
"dev": "concurrently \"npm run server\" \"npm run client\""
```

### VSCode + ESLint + Prettier

[VSCode](https://code.visualstudio.com/) is a lightweight but powerful source code editor. [ESLint](https://eslint.org/) takes care of the code-quality. [Prettier](https://prettier.io/) takes care of all the formatting.

#### Installation guide

1.  Install [VSCode](https://code.visualstudio.com/)
2.  Install [ESLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)
3.  Install [Prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
4.  Modify the VSCode user settings to add below configuration

    ```javascript
    "eslint.alwaysShowStatus": true,
    "eslint.autoFixOnSave": true,
    "editor.formatOnSave": true,
    "prettier.eslintIntegration": true
    ```

Above, we have modified editor configurations. Alternatively, this can be configured at the project level by following [this article](https://medium.com/@netczuk/your-last-eslint-config-9e35bace2f99).


google-cloud-sdk/RELEASE_NOTES
# Google Cloud SDK - Release Notes

Copyright 2014-2020 Google Inc. All rights reserved.

Subscribe to these release notes at [https://groups.google.com/forum/#!forum/google-cloud-sdk-announce](https://groups.google.com/forum/#!forum/google-cloud-sdk-announce).


## 289.0.0 (2020-04-14)

### App Engine

* Deploying Maven projects using a pom.xml file is now GA.
* Updated the Python SDK to version 1.9.90.
  Please visit the following release notes for details:
  <https://cloud.google.com/appengine/docs/python/release-notes>.

### Cloud Access Context Manager

* Added `--custom-level-spec` argument to enable custom access level definition to the following commands:
  * `gcloud access-context-manager levels create`
  * `gcloud access-context-manager levels update`
* Added `gcloud beta access-context-manager perimeters dry-run enforce` command
  for enforcing the dry-run configurations for individual Service Perimeters.
* Promoted `gcloud access-context-manager levels replace-all` and `gcloud access-context-manager perimeters replace-all` commands to GA.
* Added `gcloud beta access-context-manager perimeters dry-run enforce` command
  for enforcing the dry-run configurations for individual Service Perimeters.

### Cloud Dataproc

* Fixed bug where the GKE cluster name is also displayed in the zone column when
  listing GKE-based Dataproc clusters.

### Cloud Memorystore

*   Added new `redis_5_0` redis-version value to `gcloud redis instances
    upgrade`.

### Cloud Run

* Promoted the Knative API version used from v1alpha1 to v1 on all `gcloud beta
  run` and `gcloud run` commands for `--platform=gke` and `--platform=kubernetes`.

### Compute Engine

* Promoted `gcloud compute instances add-resource-policies` command to GA.

### Kubernetes Engine

* Promoted the following flags to GA for `gcloud container binauthz attestors
  public-keys add`:
  * `--pkix-public-key-file`
  * `--pkix-public-key-algorithm`
  * `--keyversion`
  * `--public-key-id-override`


Subscribe to these release notes at [https://groups.google.com/forum/#!forum/google-cloud-sdk-announce](https://groups.google.com/forum/#!forum/google-cloud-sdk-announce).


## 288.0.0 (2020-04-07)

### Breaking Changes

*   **(Secret Manager)** Removed newline from the end of the output of `gcloud secrets versions access` command.


### AI Platform

* Added support for `--region` flag for `gcloud beta ai-platform`.

### Cloud Access Context Manager

* Renamed `gcloud beta access-context-manager perimeters dry-run reset` as
  `gcloud beta access-context-manager perimeters dry-run drop`.

### Cloud Composer

* Added the following flags to `gcloud composer environments create` to handle
IP aliasing:
  * `--enable-ip-alias`
  * `--cluster-ipv4-cidr`
  * `--services-ipv4-cidr`
  * `--cluster-secondary-range-name`
  * `--services-secondary-range-name`

* Added the following flags to `gcloud composer environments create` to handle
Private IP:
  * `--enable-private-environment`
  * `--enable-private-endpoint`
  * `--master-ipv4-cidr`

* Added the following flags to `gcloud composer environments create` to handle
Private IP:
  * `--web-server-ipv4-cidr`
  * `--cloud-sql-ipv4-cidr`

* Fixed a bug that could cause flags passed to `gcloud composer environments
run` to be interpreted incorrectly.

### Cloud Dataflow

* Added `gcloud beta dataflow snapshots` command group which provides the ability to create and manage Cloud Dataflow snapshots.

### Cloud Dataproc

* Fixed bug where the default project was not automatically used as the GKE
  cluster project for GKE-based Dataproc clusters.
* Changed the default value for `--timeout` in `gcloud dataproc clusters update`
  from 3h to 25h. This prevents client-side timeouts during long graceful
  decommissioning operations.

### Cloud Run

* Promoted `--no-traffic` flag of `gcloud run deploy` to GA for `--platform=managed`.

### Compute Engine

* Promoted `--scale-in-control` flag of `gcloud compute instance-groups managed set-autoscaling` to beta.
* Promoted `--scale-in-control` and `--clear-scale-in-control` flags of `gcloud compute instance-groups managed update-autoscaling` to beta.
* Added support of tabular display format of command
  `gcloud compute networks get-effective-firewalls` and
  `gcloud compute instances network-interfaces get-effective-firewalls`.

### Kpt

* Updated kpt from v0.14.0 to v0.17.0.
* Updated import of cli-utils from v0.5.0 to v0.6.0.
* Updated import of kyaml from v0.1.3 to v0.1.4.
* Updated import of cmd/config from v0.1.3 to v0.1.4.


Subscribe to these release notes at [https://groups.google.com/forum/#!forum/google-cloud-sdk-announce](https://groups.google.com/forum/#!forum/google-cloud-sdk-announce).


## 287.0.0 (2020-04-01)

### App Engine

* Updated the Python SDK to version 1.9.89.
  Please visit the following release notes for details:
  <https://cloud.google.com/appengine/docs/python/release-notes>.

### BigQuery

* Added support to update capacity_commitment plan and renewal_plan.
* Modified DTS CreateTransferConfig and UpdateTransferConfig CLI to accept a
  version_info instead of an authorization_code.
* Added support for ROW ACCESS POLICY DDL statements.
* Fixed a bug with None auth_info in UpdateTransferConfig.

### Cloud Access Context Manager

* Added the following flags to `gcloud beta access-context-manager perimeters create`:
  * `--enable-vpc-accessible-services`
  * `--vpc-allowed-services`
* Added the following flags to `gcloud beta access-context-manager perimeters update`:
  * `--enable-vpc-accessible-services`
  * `--add-vpc-allowed-services`
  * `--clear-vpc-allowed-services`
  * `--remove-vpc-allowed-services`

### Cloud Asset Inventory

* Support `org-policy` and `access-policy` in the real-time feed API.

### Cloud Composer

* Added the following flags to `gcloud beta composer environments` commands to
specify IP ranges for Web Server and Cloud SQL network subranges:
  * `--web-server-ipv4-cidr`
  * `--cloud-sql-ipv4-cidr`

webpack.config.js
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const CleanWebpackPlugin = require('clean-webpack-plugin');

const outputDirectory = 'dist';

module.exports = {
  entry: ['babel-polyfill', './src/client/index.js'],
  output: {
    path: path.join(__dirname, outputDirectory),
    filename: 'bundle.js'
  },
  module: {
    rules: [{
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader'
        }
      },
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader']
      },
      {
        test: /\.(png|woff|woff2|eot|ttf|svg)$/,
        loader: 'url-loader?limit=100000'
      },
      {
        test: /\.(png|jp(e*)g|svg|gif)$/,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: 'images/[hash]-[name].[ext]',
            },
          },
        ],
      }
    ]
  },
  resolve: {
    extensions: ['*', '.js', '.jsx']
  },
  devServer: {
    port: 3000,
    open: true,
    proxy: {
      '/api': 'http://localhost:8080'
    }
  },
  plugins: [
    new CleanWebpackPlugin([outputDirectory]),
    new HtmlWebpackPlugin({
      template: './public/index.html',
      favicon: './public/favicon.ico'
    })
  ]
};